openshift-kube-apiserver/filters/deprecatedapirequest/controller.go

```diff
@@ -0,0 +1,145 @@
+package deprecatedapirequest
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	apiclientv1 "github.com/openshift/client-go/apiserver/clientset/versioned/typed/apiserver/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers"
+)
+
+// NewController returns a controller
+func NewController(config *rest.Config) *controller {
+	ret := &controller{
+		client:       apiclientv1.NewForConfigOrDie(configFor(config)).DeprecatedAPIRequests(),
+		nodeName:     nodeFor(),
+		updatePeriod: 5 * time.Minute,
+	}
+	ret.resetRequestCount()
+	return ret
+}
+
+func configFor(config *rest.Config) *rest.Config {
+	c := rest.CopyConfig(config)
+	c.AcceptContentTypes = "application/json"
+	c.ContentType = "application/json"
+	return c
+}
+
+func nodeFor() string {
+	node := os.Getenv("HOST_IP")
+	if hostname, err := os.Hostname(); err != nil {
+		node = hostname
+	}
+	return node
+}
+
+// APIRequestLogger support logging deprecated API requests.
+type APIRequestLogger interface {
+	IsDeprecated(resource, version, group string) bool
+	LogRequest(resource schema.GroupVersionResource, timestamp time.Time, user, verb string)
+	Start(stop <-chan struct{})
+}
+
+type controller struct {
+	client       apiclientv1.DeprecatedAPIRequestInterface
+	nodeName     string
+	updatePeriod time.Duration
+
+	requestCountLock sync.Mutex
+	requestCounts    *apiRequestCounts
+}
+
+// requestRecord represents an api resource request to be logged
+type requestRecord struct {
+	resource  schema.GroupVersionResource
+	user      string
+	verb      string
+	timestamp time.Time
+}
+
+// IsDeprecated return true if the resource is deprecated.
+func (c *controller) IsDeprecated(resource, version, group string) bool {
+	_, ok := deprecatedApiRemovedRelease[schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resource,
+	}]
+	return ok
+}
+
+// LogRequest queues an api request for logging
+func (c *controller) LogRequest(resource schema.GroupVersionResource, timestamp time.Time, user, verb string) {
+	c.requestCounts.IncrementRequestCount(resource, timestamp.Hour(), user, verb, 1)
+}
+
+// resetCount returns the current count and creates a new requestCount instance var
+func (c *controller) resetRequestCount() *apiRequestCounts {
+	c.requestCountLock.Lock()
```

Comment:
Don't see the need for this lock as written. Only `NewController()` and serialized access in the "write out logs" goroutine in `Start()` preform this reset. I suspect maybe you meant to use lock in `LogRequest()` ?

Reply:
> Don't see the need for this lock as written. Only `NewController()` and serialized access in the "write out logs" goroutine in `Start()` preform this reset. I suspect maybe you meant to use lock in `LogRequest()` ?

it's to prevent that race you found.